src/jxl.imageio/jxloutput.cpp

```diff
@@ -0,0 +1,338 @@
+                       JXL_NATIVE_ENDIAN, 0 };
+
+    const size_t pixels_size = m_basic_info.xsize * m_basic_info.ysize
+                               * m_basic_info.num_color_channels;
+
+    m_pixels.resize(pixels_size);
+
+    const void* data = m_pixels.data();
+    size_t size      = m_pixels.size() * sizeof(float);
+
+    DBG std::cout << "data = " << data << " size = " << size << "\n";
+
+    status = JxlEncoderAddImageFrame(m_frame_settings, &m_pixel_format, data,
+                                     size);
+    DBG std::cout << "status = " << status << "\n";
+    if (status != JXL_ENC_SUCCESS) {
+        DBG std::cout << "JxlEncoderAddImageFrame failed.\n";
+        return false;
+    }
+
+    // No more image frames nor metadata boxes to add
+    DBG std::cout << "calling JxlEncoderCloseInput()\n";
+    JxlEncoderCloseInput(m_encoder.get());
+
+    compressed.clear();
+    compressed.resize(4096);
+    uint8_t* next_out = compressed.data();
+    size_t avail_out  = compressed.size() - (next_out - compressed.data());
+    JxlEncoderStatus result = JXL_ENC_NEED_MORE_OUTPUT;
+    while (result == JXL_ENC_NEED_MORE_OUTPUT) {
+        DBG std::cout << "calling JxlEncoderProcessOutput()\n";
+        result = JxlEncoderProcessOutput(m_encoder.get(), &next_out,
+                                         &avail_out);
+        DBG std::cout << "result = " << result << "\n";
+        if (result == JXL_ENC_NEED_MORE_OUTPUT) {
+            size_t offset = next_out - compressed.data();
+            compressed.resize(compressed.size() * 2);
+            next_out  = compressed.data() + offset;
+            avail_out = compressed.size() - offset;
+        }
+    }
+    compressed.resize(next_out - compressed.data());
+    if (result != JXL_ENC_SUCCESS) {
+        DBG std::cout << "JxlEncoderProcessOutput failed.\n";
+        return false;
+    }
+
+    DBG std::cout << "compressed.size() = " << compressed.size() << "\n";
+
+    if (!iowrite(compressed.data(), 1, compressed.size())) {
+        DBG std::cout << "iowrite failed.\n";
+        return false;
+    }
+
+    DBG std::cout << "JxlOutput::save_image() return ok\n";
+    return ok;
+}
+
+bool
+JxlOutput::close()
+{
+    bool ok = true;
+
+    DBG std::cout << "JxlOutput::close()\n";
+
+    if (!ioproxy_opened()) {  // Already closed
+        init();
+        return true;
+    }
+
+    if (m_spec.tile_width) {
+        // Handle tile emulation -- output the buffered pixels
+        OIIO_ASSERT(m_tilebuffer.size());
+        ok &= write_scanlines(m_spec.y, m_spec.y + m_spec.height, 0,
+                              m_spec.format, &m_tilebuffer[0]);
+        std::vector<unsigned char>().swap(m_tilebuffer);
+    }
+
+    save_image();
+
+    init();
+    return ok;
+}
+
+bool
+JxlOutput::copy_image(ImageInput* in)
+{
+    if (in && !strcmp(in->format_name(), "jxl")) {
+        return true;
+    }
+
+    return ImageOutput::copy_image(in);
+}
```

Comment:
I don't think this is correct at all. If I understand, what you're saying is that if it's copying from another jxl file, it DOES NOTHING but returns true indicating success? If you don't have a special way to copy from jxl to jxl efficiently without decompressing and recompressing, then you don't need to provide a copy_image method at all.